apis/apiextensions/fn/proto/v1beta1/run_function.proto

```diff
@@ -173,4 +187,10 @@ enum Severity {
   // Normal results are emitted as normal events and debug logs associated
   // with the composite resource.
   SEVERITY_NORMAL = 3;
+}
+
+// The environment to be passed down the Function pipeline.
+message Environment {
+  // The data section of the environment.
+  google.protobuf.Struct data = 1;
```

Comment:
Will Functions use keys in this `data` struct to pass information to each other?

My understanding from https://github.com/crossplane/crossplane/pull/4731#issuecomment-1748183984 is "no". If that's correct, I think it would be useful to bring that use case into scope for this PR. Would we add a new field to the `Environment` message for that?

Reply:
I could imagine foregoing the `Environment` message and just having something like:

```proto
message RunFunctionRequest {

  // Fields omitted for brevity

  google.protobuf.Struct environment = 5;
}
```

You could then put the merged `EnvironmentConfigs` under a well-known key like `apiextensions.crossplane.io/config` and let Functions use other fields within the `environment` struct to pass information without worrying about conflicting with what is loaded from the `EnvironmentConfigs`.

Reply:
> Will Functions use keys in this data struct to pass information to each other?
What do you mean? "data" is what we get today merging EnvironmentConfig and patching to and from the environment, if we were to add other fields to the environment these would be on the same level as the "data".

`environment.data` is exactly what you were putting at `environment[apiextensions.crossplane.io/environment]`. I would let people use FromEnvironment patches adding the "data" prefix to fields, so that if we ever added something in the future it could still be reachable by FromEnvironmentPatches in the patch and transform function.

Reply:
> I would let people use FromEnvironment patches adding the "data" prefix to fields

We're on the same page here pattern wise, i.e. nesting what we would call "the environment" today under a key in... the environment. I think we need a more descriptive, specific names than "environment" and "data" though. If my EnvironmentConfigs and defaults become the "data" key in the environment, what else would there be? Would the other "keys" be something other than data?

Perhaps if we really want to stick with "environment" as the top-level name `environment.data` could become `environment.config`? or `environment[apiextensions.crossplane.io/config]`?

> If we ever added something in the future it could still be reachable by FromEnvironmentPatches in the patch and transform function.

I feel that v1.14 we should have a story for Functions to write to the "environment" for v1.14. If we merge this PR as-is it seems like their only option will be to write to `environment.data`, which means they could potentially collide with the object that Crossplane loads into that struct from EnvConfigs.

I lean toward the top level of the "environment" being a struct or map because then a Function can "claim" an arbitrary field/key within that top-level struct/map without needing an update to the protobuf schema to add more fields.

Reply:
I still don't feel great about having a well-known key we use "officially" to inject the current environment into an unstructured object.

I'd be more comfortable saying that the environment (or context) is just an unstructured object and letting functions define their convention, e.g. an EnvironmentConfig aggregating function defines a key it will put the aggregated result in some key, which will be the same key expected by the patch and transform function, ignoring the whole spec.environment section in Pipeline mode. This was what I was hinting at here: https://github.com/crossplane/crossplane/pull/4632#discussion_r1346177527.

My point is, if stuff is official it should be explicitly represented in the schema, otherwise let's fully go the other way and let functions define their own conventions, just providing them a way to pass data through the pipeline outside the desired state.

Reply:
Maybe a summary of my thoughts will help me communicate clearly. 😄 

* I get why we want to stick with "environment" as the top-level name. I'm on board with that.
* I'm on board with what we load from EnvConfigs ("data") being a field or key in the top-level environment.
* I think we need to find a more descriptive name than "data", though. [^1]
* I think Functions should be able to create arbitrary fields/keys in the top-level environment.
* I'm on board with `PatchFromEnvironment` fields being "rooted" at the top-level object. So it can access both "data" and whatever Functions write to it.

[^1]: Naming is important to me here because it will help bridge how folks think about "the environment". Today I think it means "stuff loaded from EnvConfigs" (+defaults). Here it sounds like "the environment" becomes one layer higher than that.

Reply:
> I think we need to find a more descriptive name than "data", though. 

If we do go with a map / arbitrary keys, I think it makes sense to reserve and use certain key namespaces for Crossplane (core, not Functions) to write to. Similar to how `kubernetes.io/` and `k8s.io/` annotation key prefixes are [reserved for use by Kubernetes components](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/#syntax-and-character-set), and `GITHUB_` env vars are [reserved for use by GitHub in GitHub Workflows](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-environment-variable).

Reply:
> I still don't feel great about having a well-known key we use "officially" to inject the current environment into an unstructured object.

Sorry, GitHub didn't show me this comment until I refreshed the page.

I think I get your reservations - explicit schemas are usually best. That said, your analogy of the environment being like environment variables, and looking at how other systems use environment variables is making me feel better about the pattern.

If we say that Functions can effectively communicate by reading from and writing to shared "environment variables", i.e. top-level keys in the environment, is it so bad if Crossplane "bootstraps" some of that environment, like GitHub workflows does with their env vars?

Reply:
Yeah, I don't feel super strongly against that and I can see you point, so I would be fine with annotations-style reserved keyspaces. I'll revert the whole environment to be just a struct tomorrow. Now the only point is whether we want to wire the current EnvironmentConfig retrieval from configurations or just accept that people are going to access the API server (or any other external system) from functions and just deal with it, which is the topic of the other thread above, so let's continue there if you agree.

Reply:
done: https://github.com/crossplane/crossplane/pull/4632/commits/2b28d9a136c6bc33352039596250c9c5923b5779

Reply:
called it `context` as you were doing in #4731, as it didn't make much sense to put the current environment in the environment but under a key `.../environment`...